File: polb/cnvrt.f
      subroutine cnvrt(imode,ain1,ain2,out1,out2,it,alat,wlong)
C
C  INPUT:
C
C     IMODE - specifies input and output coordinate systems
C              IMODE     Input      Output
C              -----     -----      ------
C                1       RA/DEC      AZ/EL
C                2       AZ/EL       RA/DEC
C                3       RA/DEC      X/Y NS
C                4       X/Y NS      AZ/EL
C                5       AZ/EL       X/Y NS
C                6       X/Y NS      HA/DEC
C                7       X/Y NS      RA/DEC
C                8       HA/DEC      X/Y NS
C                9       HA/DEC      AZ/EL
C               10       AZ/EL       HA/DEC
C
C     AIN1 - first input coordinate
C     AIN2 - second input coordinate
C
C  OUTPUT:
C
C     OUT1 - first output coordinate
C     OUT2 - second output coordinate
C
      implicit double precision (a-h,o-z)
      double precision sidt,sider,had,tlst
      real dut
      dimension it(6)
C
      include '../include/dpi.i'
C
      dasin(x) = datan2z(x,dsqrt(dabs(1.-x*x)))
      dacos(x)=datan2z(dsqrt(dabs(1.-x*x)),x)
C
      sidt=sider(it)
      goto 10
C
      entry cnvrt2(imode,ain1,ain2,out1,out2,it,dut,alat,wlong)
      sidt=sider2(it,dut)
C
 10   continue
      slat = dsin(alat)
      clat = dcos(alat)
      sin1 = dsin(ain1)
      sin2 = dsin(ain2)
      cin1 = dcos(ain1)
      cin2 = dcos(ain2)
      tlst=sidt-wlong
      had = tlst-ain1
      if (had.gt.0.d0) had=dmod(had,dtwopi)
      if (had.lt.0.d0) had=dmod(had,-dtwopi)
      if (had.lt.-dpi) had=had+dtwopi
      if (had.gt.dpi) had=had-dtwopi
      ha=had
      sha=dsin(ha)
      cha=dcos(ha)
C
      goto (110,130,150,170,190,210,230,250,270,290) imode
C
C  1. RA/DEC --> AZ/EL
C
110   out2=dasin(slat*sin2+clat*cin2*cha)
      out1=datan2z(-cin2*sha,clat*sin2-slat*cin2*cha)
      if (out1.lt.0.) out1=out1+dtwopi
      goto 900
C
C  2. AZ/EL --> RA/DEC
C
130   ha=-datan2z(cin2*sin1,sin2*clat-cin2*cin1*slat)
      out1=tlst-ha
      if (out1.lt.0.) out1=out1+dtwopi
      if (out1.gt.dtwopi) out1=out1-dtwopi
      out2=dasin(cin2*cin1*clat+sin2*slat)
      goto 900
C
C  3. RA/DEC --> X/Y NS
C
150   out1=datan2z(-cin2*sha,slat*sin2+clat*cin2*cha)
      out2=dasin(clat*sin2-slat*cin2*cha)
      goto 900
C
C  4. X/Y NS --> AZ/EL
C
170   out1=datan2z(sin1*cin2,sin2)
      if (out1.lt.0) out1=out1+dtwopi
      out2=dasin(cin2*cin1)
      goto 900
C
C  5. AZ/EL  --> XY NS
C
190   continue
      out1=datan2z(cin2*sin1,sin2)
      out2=dasin(cin2*cin1)
      goto 900
C
C  6. X/Y  NS --> HA/DEC
C
210   continue
      out1=-datan2z(cin2*sin1,cin2*cin1*clat-sin2*slat)
      out2=dasin(sin2*clat+cin2*cin1*slat)
      goto 900
C 
C   7. X/Y NS --> RA/DEC  
C
230   continue
      ha=-datan2z(cin2*sin1,cin2*cin1*clat-sin2*slat)
      out1=tlst-ha
      if (out1.lt.0.0) out1=out1+dtwopi
      if (out1.gt.dtwopi) out1=out1-dtwopi
      out2=dasin(sin2*clat+cin2*cin1*slat)
      goto 900
C
C  8. HA/DEC --> X/Y NS
C
250   out1=datan2z(-cin2*sin1,slat*sin2+clat*cin2*cin1)
      out2=dasin(clat*sin2-slat*cin2*cin1)
      goto 900
C
C  9. HA/DEC --> AZ/EL
C
270   out2=dasin(slat*sin2+clat*cin2*cin1)
      out1=datan2z(-cin2*sin1,clat*sin2-slat*cin2*cin1)
      if (out1.lt.0.) out1=out1+dtwopi
      goto 900
C
C 10. AZ/EL --> HA/DEC
C
290   out1=-datan2z(cin2*sin1,sin2*clat-cin2*cin1*slat)
      if (out1.lt.-dpi) out1=out1+dtwopi
      if (out1.gt.dpi) out1=out1-dtwopi
      out2=dasin(cin2*cin1*clat+sin2*slat)
      goto 900
C
900   continue

      return
      end

File: monit/cnvrt.c
/* cnvrt

   Convert from one coordinate system to another, depending
   on the input mode. All angle arguments are radians.
   Copied from the FORTRAN.
   NRV 920311

   imode     Input      Output
   -----    -------    --------
     1      ra/dec     az/el
     2      az/el      ra/dec
     3      ra/dec     x/y NS
     4      x/y NS     az/el
     5      az/el      x/y NS
     6      x/y NS     ha/dec
     7      x/y NS     ra/dec
     8      ha/dec     x/y NS
     9      ha/dec     az/el
    10      az/el      ha/dec
*/

#include <math.h>
#include <stdio.h>
#include "dpi.h"
#include "stparams.h"

void sider();

cnvrt(mode,ain1,ain2,out1,out2,it,alat,wlong)
int mode;               /* type of conversion */
double ain1, ain2;      /* input angles       */
double *out1, *out2;    /* output angles      */
int it[6];              /* standard rte time  */
double alat,wlong;      /* lat,lon of station */

{
  double ha,sidt,tlst;
  double slat,clat,sin1,sin2,cin1,cin2,sha,cha;
/*
  fprintf(stdout,"cnvrt: mode=%d, in=%f,%f\n",
  mode,ain1*RAD2DEG,ain2*RAD2DEG);
  fprintf(stdout,"   it=%d/%d-%d:%d:%d\n",it[5],it[4],it[3],it[2],it[1]);
*/
  slat = sin(alat);
  clat = cos(alat);
  sin1 = sin(ain1);
  sin2 = sin(ain2);
  cin1 = cos(ain1);
  cin2 = cos(ain2);
  sider(it,it[5],&sidt);
/*
  fprintf(stdout,"sidt= %f\n",sidt*RAD2DEG);
*/
  tlst = sidt - wlong;
  ha = tlst - ain1;
/*
  fprintf(stdout,"ha= %f\n",ha*RAD2DEG);
*/
  ha = (ha >  0.0) ? fmod(ha,DTWOPI) : fmod(ha,-DTWOPI);
  ha = (ha < -DPI) ? ha+DTWOPI    : ha;
  ha = (ha >  DPI) ? ha-DTWOPI    : ha;
  sha = sin(ha);
  cha = cos(ha);

  switch (mode) {

    case 1:               /* ra/dec --> az/el */
      *out2 = asin(slat*sin2 + clat*cin2*cha);
      *out1 = ATAN2Z(-cin2*sha,clat*sin2-slat*cin2*cha);
      if (*out1 < 0.0)
        *out1 += DTWOPI;
      break;

    case 2:               /* az/el --> ra/dec */
      ha = -ATAN2Z(cin2*sin1,sin2*clat-cin2*cin1*slat);
      *out1=tlst-ha;
      if (*out1 < 0.0) *out1 += DTWOPI;
      if (*out1 > DTWOPI) *out1 -= DTWOPI;
      *out2 = asin(cin2*cin1*clat+sin2*slat);
      break;

    case 3:               /* ra/dec --> x/y NS */
      *out1 = ATAN2Z(-cin2*sha,slat*sin2+clat*cin2*cha);
      *out2 = asin(clat*sin2-slat*cin2*cha);
      break;

    case 4:               /* x/y NS --> az/el  */
      *out1 = ATAN2Z(sin1*cin2,sin2);
      if (*out1 < 0.0) *out1 += DTWOPI;
      *out2 = asin(cin2*cin1);
      break;

    case 5:               /* az/el --> x/y NS  */
      *out1 = ATAN2Z(cin2*sin1,sin2);
      *out2 = asin(cin2*cin1);
      break;

    case 6:               /* x/y NS --> ha/dec */
      *out1 = -ATAN2Z(cin2*sin1,cin2*cin1*clat-sin2*slat);
      *out2 = asin(sin2*clat+cin2*cin1*slat);
      break;

    case 7:               /* x/y NS --> ra/dec */
      ha = -ATAN2Z(cin2*sin1,cin2*cin1*clat-sin2*slat);
      *out1=tlst-ha;
      if (*out1 < 0.0) *out1 += DTWOPI;
      if (*out1 > DTWOPI) *out1 -= DTWOPI;
      *out2 = asin(sin2*clat+cin2*cin1*slat);
      break;

    case 8:               /* ha/dec --> x/y NS */
      *out1 = ATAN2Z(-cin2*sin1,slat*sin2+clat*cin2*cin1);
      *out2 = asin(clat*sin2-slat*cin2*cin1);
      break;

    case 9:               /* ha/dec --> az/el  */
      *out2 = asin(slat*sin2+clat*cin2*cin1);
      *out1 = ATAN2Z(-cin2*sin1,clat*sin2-slat*cin2*cin1);
      if (*out1 < 0.0) *out1 += DTWOPI;
      break;

    case 10:              /* az/el --> ha/dec  */
      *out1 = ATAN2Z(cin2*sin1,sin2*clat-cin2*cin1*slat);
      if (*out1 < -DPI) *out1 += DTWOPI;
      if (*out1 >  DPI) *out1 -= DTWOPI;
      *out2 = asin(cin2*cin1*clat+sin2*slat);
  }
/*
  printf("cnvrt: out=%f %f\n",*out1*RAD2DEG,*out2*RAD2DEG);
*/
}
